File: apps/web/lib/reconciliation.ts
/*
 * True for an expense / income row that a reconciliation created — the adjustment posted by the card
 * or the account reconciliation flow. Such a row is derived, not authored: its amount IS the
 * reconciliation's recorded `difference`, so mutating it directly makes the reconciliation lie.
 *
 * Both the Edit and the Delete affordance are withheld, because the API refuses both with
 * 409 `reconciliation_owned_entry`. Delete is not an escape hatch: the reverse pointer
 * (`card_reconciliations/account_reconciliations.adjustment_*_id`) is ON DELETE SET NULL, so removing
 * the entry leaves the reconciliation alive with a null pointer and a `difference` it no longer
 * applies, while the balance silently snaps back.
 *
 * The supported way to change one is to re-run or delete its RECONCILIATION, which recomputes or
 * cascade-drops the adjustment (`expense_entries/income_entries.{reconciliation_id,
 * account_reconciliation_id}` are ON DELETE CASCADE).
 *
 * Keyed on the foreign keys, not on `source === 'reconciliation'`, so this predicate agrees exactly
 * with the backend guard. `source` records provenance and survives a restore, which nulls both links —
 * a restored adjustment is a plain historical entry that nothing owns, so it stays mutable. What such
 * a row still cannot do is round-trip the entry form, and that is a separate concern owned by
 * isSystemExpenseCategory / isSystemIncomeCategory in lib/utils/categories.
 */
export function isReconciliationOwned(entry: {
  reconciliationId: number | null;
  accountReconciliationId: number | null;
}): boolean {
  return entry.reconciliationId !== null || entry.accountReconciliationId !== null;
}

File: apps/api/app/domain/reconciliation.py
# Reconciliation rules shared by the expense and income services and their request schemas. Both
# reconciliation features — card (Phase 3, Step 5) and account (the cash/bank money engine) — post
# their difference as an ordinary expense_entries / income_entries row linked back to the
# reconciliation that created it, categorised with a value reserved for true-ups. This module owns the
# two rules that keep such a row honest: only a reconciliation may own one, and only a reconciliation
# may write its category.

from app.domain.errors import ReconciliationOwnedEntryError
from app.models.expense_entry import ExpenseCategory
from app.models.income_entry import IncomeCategory

# Category values only a reconciliation writes — they LABEL a true-up, which is what lets the app
# separate real spending from a balance correction. A user-supplied value here would be a fake
# true-up: indistinguishable in every breakdown from one the app actually computed. Neither picker
# offers them, and the request schemas reject them (see app/schemas/base.py), so they can only ever
# arrive from the reconciliation services, which build their rows through the repositories directly.
SYSTEM_EXPENSE_CATEGORIES: frozenset[ExpenseCategory] = frozenset(
    {
        ExpenseCategory.account_adjustment,
        ExpenseCategory.card_credits_and_refunds,
        ExpenseCategory.card_fees_and_taxes,
    }
)

# The income counterpart. card_credits_and_refunds is legacy — card credits are signed expenses now,
# so nothing writes it — but it stays reserved rather than becoming user-pickable, because historical
# rows carry it and it would otherwise start meaning two different things.
SYSTEM_INCOME_CATEGORIES: frozenset[IncomeCategory] = frozenset(
    {
        IncomeCategory.account_adjustment,
        IncomeCategory.card_credits_and_refunds,
    }
)


# Raises ReconciliationOwnedEntryError when either reconciliation link is set, i.e. the row is a
# reconciliation's adjustment rather than an entry the user authored. Called by the expense / income
# services on update and delete, before any mutation or side-effect lookup, so a rejected request
# writes nothing. Both links are checked because the two features use separate columns:
# reconciliation_id (card) and account_reconciliation_id (account); a row carries at most one.
#
# Keyed on the foreign keys rather than source == 'reconciliation' deliberately. `source` records
# provenance and survives a restore, which nulls both links — a restored adjustment is a plain
# historical entry that nothing owns, so it must stay mutable. Other flows (scheduler, importers,
# shortcut) also write `source`, so it is the wrong fact to gate on.
def ensure_not_reconciliation_owned(reconciliation_id: int | None, account_reconciliation_id: int | None) -> None:
    if reconciliation_id is not None or account_reconciliation_id is not None:
        raise ReconciliationOwnedEntryError()
